scipy/spatial/transform/_rotation.pyx

```diff
@@ -223,6 +236,140 @@ cdef double[:, :] _compute_euler_from_matrix(
+):
+    # The algorithm assumes extrinsic frame transformations. The algorithm
+    # in the paper is formulated for rotation quaternions, which are stored
+    # directly by Rotation.
+    # Adapt the algorithm for our case by
+    # 1. Replacing "d" with "-d" for intrinsic rotation
+    # 2. Reversing both axis sequence and angles for extrinsic rotations
+    
+    if not extrinsic:
+        seq = seq[::-1]
+
+    cdef int i = _elementary_basis_index(seq[0])
+    cdef int j = _elementary_basis_index(seq[1])
+    cdef int k = _elementary_basis_index(seq[2])
+    
+    # quick renormalization of transformed quaternion, assumes |q| = 1
+    cdef int norm_squared 
+
+    is_proper = i == k
+    if is_proper:
+        k = 3 - i - j # get third axis
+        norm_squared = 1
+    else:
+        norm_squared = 2
+        
+    # Step 0
+    # Check if permutation is even (+1) or odd (-1)     
+    cdef int sign = (i-j)*(j-k)*(k-i)//2
+
+    cdef Py_ssize_t num_rotations = quat.shape[0]
+
+    # some forward definitions
+    cdef double[:, :] angles = _empty2(num_rotations, 3)
+    cdef double[:] _angles # accessor for each rotation
+    cdef double a, b, c, d, cos_theta_2
+    cdef double half_sum, half_diff
+    cdef double eps = 1e-7
+    cdef bint safe1, safe2, safe, adjust
+
+    for ind in range(num_rotations):
+        _angles = angles[ind, :]
+
+        # Step 1
+        # Permutate quaternion elements            
+        if is_proper:
+            a = quat[ind, 3]
+            b = quat[ind, i]
+            c = quat[ind, j]
+            d = quat[ind, k] * sign
+        else:
+            a = quat[ind, 3] - quat[ind, j]
+            b = quat[ind, i] + quat[ind, k] * sign
+            c = quat[ind, j] + quat[ind, 3]
+            d = quat[ind, k] * sign - quat[ind, i]
+        
+        # Step 2
+        # Ensure less than unit norm
+        cos_theta_2 = 2*(a**2 + b**2)/norm_squared - 1
+        cos_theta_2 = min(cos_theta_2, 1)
+        cos_theta_2 = max(cos_theta_2, -1)
+        
+        # Step 3
+        # Compute second angle...
+        _angles[1] = acos(cos_theta_2)
+
+        # ... and check if equal to is 0 or pi, causing a singularity
+        safe1 = abs(_angles[1]) >= eps
+        safe2 = abs(_angles[1] - <double>pi) >= eps
+        safe = safe1 and safe2
+
+        # Step 4
+        # compute first and third angles, according to case
+        if safe:
+            half_sum = atan2(b, a) # == (angles[0] + angles[2])/2
+            half_diff = atan2(-d, c) # == (angles[0] - angles[2])/2 
+        
+            _angles[0] = half_sum + half_diff
+            _angles[2] = half_sum - half_diff
+
+        if not extrinsic:
+            # For intrinsic, set first angle to zero so that after reversal we
+            # ensure that third angle is zero
+            # 
+            if not safe:
+                _angles[0] = 0
+            # 
+            if not safe1:
+                half_sum = atan2(b, a)
```

Comment:
Looks like we can assign `_angles[2] = 2 * atan2(b, a)` without updating `half_sum`. Looks slightly cleaner in my opinion.

Reply:
I will change it for now, but I had decided (maybe unnecessarily) to separate it in order to more easily visualize a future option I planned on proposing:

I briefly talk about it in the paper too, basically: in case of singularities, either the equation for `half_sum` or `half_diff` is undetermined because: we have either one or the other, and the representation is **not unique**.

What we do by now is: when this happens, we assign `_angles[0] = 0` and use this with the remaining equation to calculate `_angles[2]` (or `_angles[2] = 0` and we calculate `_angles[0]`).

What I would like to propose as an option with a special argument is: instead of assuming one of the angles is zero, we could actually take into advantage this free choice and assign it to the the previous value, keeping it constant: `_angles[0] = angles[ind-1, 0]`. This way, the new value calculated for `_angles[2]` has the advantage of being **continuous**.